device/common/include/traccc/finding/device/impl/prune_tracks.ipp

```diff
@@ -0,0 +1,43 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+namespace traccc::device {
+
+TRACCC_DEVICE inline void prune_tracks(
+    std::size_t globalIndex,
+    track_candidate_container_types::const_view track_candidates_view,
+    vecmem::data::vector_view<const unsigned int> valid_indices_view,
+    track_candidate_container_types::view prune_candidates_view) {
+
+    track_candidate_container_types::const_device track_candidates(
+        track_candidates_view);
+    vecmem::device_vector<const unsigned int> valid_indices(valid_indices_view);
+    track_candidate_container_types::device prune_candidates(
+        prune_candidates_view);
+
+    if (globalIndex >= prune_candidates.size()) {
```

Comment:
Should this not rather be

```suggestion
    if (globalIndex >= valid_indices.size()) {
```

? 😕 Or are these two equivalent? (We may want to add an assertion if so.)

Reply:
Actually they don't have the same size. `valid_indices` is larger than `prune_candidates` but the number of non-trivial elements are the same.

`globalIndex` check should also be done with `prune_candidates.size()` as the thread block dimension was set based on its value.

Reply:
So, what's the logic then? :confused: I thought `valid_indices` was used to filter a larger container down to a smaller one. :thinking: Is the idea that the first N elements of `valid_indices` would describe the N elements of `prune_candidates`? :thinking:

Reply:
> Is the idea that the first N elements of valid_indices would describe the N elements of prune_candidates?

Exactly. The reason that size of valid_indices (=number of tips) is larger than N (=number of valid tips) is just because we don't know its exact size during the kernel of `build_tracks`.